barretenberg/cpp/src/barretenberg/stdlib_circuit_builders/goblin_ultra_circuit_builder.cpp

```diff
@@ -219,40 +225,39 @@ template <typename FF> void GoblinUltraCircuitBuilder_<FF>::set_goblin_ecc_op_co
 }
 
 /**
- * @brief Read from calldata
- * @details Creates a calldata lookup gate based on the read data
+ * @brief Read from a databus column
+ * @details Creates a databus lookup gate based on the input index and read result
  *
  * @tparam FF
  * @param read_idx_witness_idx Variable index of the read index
  * @return uint32_t Variable index of the result of the read
  */
-template <typename FF> uint32_t GoblinUltraCircuitBuilder_<FF>::read_calldata(const uint32_t& read_idx_witness_idx)
+template <typename FF>
+uint32_t GoblinUltraCircuitBuilder_<FF>::read_bus_vector(BusVector& bus_vector, const uint32_t& read_idx_witness_idx)
 {
-    // Get the raw index into the calldata
+    // Get the raw index into the databus column
     const uint32_t read_idx = static_cast<uint32_t>(uint256_t(this->get_variable(read_idx_witness_idx)));
 
-    // Ensure that the read index is valid
-    ASSERT(read_idx < public_calldata.size());
+    ASSERT(read_idx < bus_vector.size());            // Ensure that the read index is valid
+    ASSERT(bus_vector.get_read_count(read_idx) < 1); // Reading more than once at the same index is not supported
```

Comment:
This is done for efficiency reasons, I added an issue explaining this in detail. I marked it as NOTE rather than TODO since this may be an acceptable restriction indefinitely